pkg/issuer/acme/setup_test.go

```diff
@@ -0,0 +1,520 @@
+
+		// Private key returned by keyFromSecret stub.
+		kfsKey crypto.Signer
+		// Error returned by keyFromSecret stub.
+		kfsErr error
+		// Whether keyFromSecret should be called.
+		kfsShouldBeCalled bool
+
+		// Whether RemoveClient should be called.
+		removeClientShouldBeCalled bool
+
+		// Whether AddClient should be called.
+		addClientShouldBeCalled bool
+
+		// Error returned by cl.Register
+		registerErr error
+
+		// ACME account returned by cl.GetReg
+		getRegAcc *acmeapi.Account
+		// Error returned by cl.GetReg
+		getRegErr error
+
+		acmePrivKeySecretCreateErr error
+
+		eabSecret       *corev1.Secret
+		eabSecretGetErr error
+
+		// expected ACME account passed to cl.Register
+		expectedRegisteredAcc *acmeapi.Account
+		// expected issuer conditions after Setup has been called.
+		expectedConditions []cmapi.IssuerCondition
+		expectedEvents     []string
+		wantsErr           bool
+	}{
+		"LetsEncrypt ACME v1 prod URL specified, return early": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEURL(acmev1Prod)),
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorInvalidConfig),
+					gen.SetIssuerConditionMessage(fmt.Sprintf(messageTemplateUpdateToV2, acmev1Prod, acmev2Prod))),
+			},
+		},
+		"LetsEncrypt ACME v1 staging URL with trailing slash specified, return early": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEURL(fmt.Sprintf("%s/", acmev1Staging))),
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorInvalidConfig),
+					gen.SetIssuerConditionMessage(fmt.Sprintf(messageTemplateUpdateToV2, fmt.Sprintf("%s/", acmev1Staging), acmev2Staging))),
+			},
+		},
+		"ACME private key secret does not exist, account key generation not disabled, key secret creation fails": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEPrivKeyRef(issuerSecretKeyName)),
+			kfsShouldBeCalled:          true,
+			kfsErr:                     notFoundErr,
+			acmePrivKeySecretCreateErr: someErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountRegistrationFailed),
+					gen.SetIssuerConditionMessage(messageAccountRegistrationFailed+someErr.Error())),
+			},
+			wantsErr: true,
+		},
+		"ACME private key secret does not exist, account key generation is disabled": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEDisableAccountKeyGeneration(true),
+			),
+			kfsShouldBeCalled: true,
+			kfsErr:            notFoundErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountVerificationFailed),
+					gen.SetIssuerConditionMessage(messageAccountVerificationFailed+messageNoSecretKeyGenerationDisabled+notFoundErr.Error())),
+			},
+			wantsErr: true,
```

Comment:
Oh, one more thought....this returns an error, which will cause a retry-with-backoff....but that's not necessary since the user creation of the missing secret , or the user changing the `DisableAccountKeyGeneration` value will trigger a re-reconcile anyway.

Reply:
I see, we watch `Secret`s [here](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/issuers/controller.go#L85) (func name wrongly suggests we only watch for deletion) and [here](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/issuers/checks.go#L43) we en-queue the `Issuer` so no need to re-enqueue from `Setup` func, as you say.

We appear to make the same mistake with other `Issuer` secrets, such as the CA secret [here](https://github.com/jetstack/cert-manager/blob/master/pkg/issuer/ca/setup.go#L55).

I could either fix this for ACME account key secret in this PR by returning nil or look at all of the `Issuer` secrets in a separate PR. I wonder if this also needs to be tested somehow, we don't currently have integration tests for `Issuer`s.

In general, we have had folks reporting issues related to running `cert-manager` in clusters with large number of `Secret`s i.e #3748 - I wonder if `Secret` handling here could perhaps be improved.

Reply:
Good idea. Keep this PR focused on testing existing behaviour and then refactor in future PRs. 
I hadn't even looked at https://github.com/jetstack/cert-manager/blob/0ff2b8778c51e6cebe140a6b196e7a9a28cbee87/pkg/controller/issuers/controller.go#L85-L120 but I see what you mean about the misleading function name which actually handles OnAdd, ONUpdate, and OnDelete, AFAICS.

Reply:
But perhaps just add a TODO comment to this test  and the corresponding code explaining that returning an error in this case is wrong and link to a GitHub issue.

Reply:
Created and issue https://github.com/jetstack/cert-manager/issues/4004